### ISSUE TITLE
Idea: PostGIS support

### ISSUE THREAD

**Issue by gijs** [sasha-alias/sqltabs#61]
Since charts are already possible, wouldn't it be a great addition to draw records with a valid geom column on a (canvas/webgl/mapnik) map?

I'm sure it's a lot harder to implement than a chart, esp. with projections and things like that... but a basic version should be feasible. I'll try and build it myself if I find some spare time.

**Comment by sasha-alias** [comment]
I've never worked with PostGIS to be honest, but since there is already used an awesome d3 visualisation library in SQL Tabs, I think it's quite possible.
What I would need for that is a more detailed user story, i.e. an example of a dataset and the way it should be visualised.

**Comment by jackrsteiner** [comment]
Agree. This would be totally awesome. It would be very much like [SharpGIS's spatial query viewer](http://www.sharpgis.net/page/SqlSpatial-Query-Tool) or [Tatukgis's Viewer](https://www.tatukgis.com/Products/Editor-Viewer/Description.aspx) or [Manifold's Viewer](http://www.manifold.net/viewer.shtml), all of which do **not** work on multiple platforms.

You can see how Manifold's interface works [on youtube](https://www.youtube.com/watch?v=kLRUXp_2El8).

With sqltabs style interface, and some arbitrary spatial table, a query might look something like:
`--- map basic
SELECT
    ST_asgeojson(geom),
    name
FROM some_table
WHERE population > 10000000;`

As sqltabs is already using d3 [DataMaps](http://datamaps.github.io) might be a good library to use. I don't understand their licensing arrangement, but [Google's GeoChart](https://developers.google.com/chart/interactive/docs/gallery/geochart), seems pretty simple to work with too.

**Comment by sasha-alias** [comment]
Technically to draw map and geo objects should be possible. D3 framework supports geojson and topojson. I'm not sure how hard to write convertion from postgis to geojson/topojson in JavaScript though.
I guess the convertor should be the first step to the goal.

**Comment by jackrsteiner** [comment]
Sasha, I'd like to look more into trying this. Can you point me toward an analogous "converter" in javascript, for something other than postgis to geojson, so I can better understand what's needed?

**Comment by gijs** [comment]
@jackrsteiner I had a look at this yesterday and found out about the [renderMap()](https://github.com/sasha-alias/sqldoc/blob/master/src/SqlDoc.js#L591) function in [sasha-alias/sqldoc](https://github.com/sasha-alias/sqldoc) (which is used by sqltabs to render the results pane). 

I'd prefer to use [react-leaflet](https://github.com/PaulLeCam/react-leaflet) (since React is already part of the project), or plain Leaflet.

One thing I couldn't figure out is how to detect if the connected database supports spatial queries?

**Comment by gijs** [comment]
Thinking out loud:

Query:

```sql
--- map points
SELECT ST_AsGeoJSON(geom), name FROM some_table WHERE population > 1000000;
```

Output:

```
(100 rows)
#  geom                                               name
1  {"type": "Feature", "geometry": { type: "Point",   "Some name"
2  {"type": "Feature", "geometry": { type: "Point",   "Some name"
3  {"type": "Feature", "geometry": { type: "Point",   "Some name"
...
```

Dataset passed to react-leaflet:

```js
...
render() {
  const { data, position, zoom } = this.props;
  // 'data' contains the data from the query, and needs to be wrapped in a FeatureCollection.
  // Coordinates need to be in EPSG:4326 et cetera...
  return (
    <Map center={position} zoom={zoom}>
      <GeoJSON data={data} />
    </Map>
  );
}
...
```

**Comment by sasha-alias** [comment]
@gijs 
you are right, `renderMap` was supposed to be used for maps rendering but I didn't succeed in implementation at that time. The design was as simple as:
```
--- chart map
SELECT ...
```
I don't mind using any license compatible lib, react-leaf is MIT so if it does the job then it should be totally fine.
Unfortunately I can't help much with implementation since I don't have any experience with spatial data, but I'm here to help you with any SQLTabs internals.

> One thing I couldn't figure out is how to detect if the connected database supports spatial queries?

You probably shouldn't care about it, if DB doesn't support spatial queries it will return exception so the dataset will be different and sqldoc will render error.

If you need to test changes in sqldoc together with sqltabs I usually use the following approach:

terminal 1:
```
cd sqldoc
npm run build:watch
```

terminal 2:
```
cd sqltabs
npm run build:watch
```

terminal 3:
```
cd sqltabs
install-local ../sqldoc/ ;  npm start
```

prerequisity:
```
npm install -g install-local
```

**Comment by jackrsteiner** [comment]
@gijs I'm not sure that it makes sense to specify the geometry type after the map trigger. Geometry types are already specified in the GeoJSON object, and Leaflet parses these. But maybe I'm missing something.

I do like the idea that the `map` keyword could be followed by styling and/or tileLayer keywords -- so perhaps a basic style template/tile is defined and set as the default, and additional templates and tileLayers could be added later.

Thanks!

**Comment by jackrsteiner** [comment]
But then again, maybe you don't want to add that complexity.

**Comment by jackrsteiner** [comment]
[Leaflet-providers](https://github.com/leaflet-extras/leaflet-providers) might be valuable, if you want to make different providers available.

**Comment by sasha-alias** [comment]
I've never saw it, but try the above fix

**Comment by jackrsteiner** [comment]
I've made a [pull request](https://github.com/sasha-alias/sqldoc/pull/2) in sqldoc. There was one other example of a variable not scoped with var, so I just added the keyword. Now following the above steps works.
